Credential Providers Refactoring

liquidz00 opened the pull request:
## Overview
This PR implements the following:

- Refactored `BasicAuthCredentialProvider` to `UserCredentialsProvider` throughout codebase
- Refactored `LoadFromAWSSecretsManager`, `PromptForCredentials` and `LoadFromKeychain` into helper functions instead of classes. Each function returns a `CredentialProvider` type that is specified or raises a `TypeError` if an invalid credential provider was passed. 
- Adjusted top-level imports of package accordingly--credential provider classes and helper functions can be imported at the top level
- Updated project docs with refactors, breaking changes, and added section on helper functions in the getting started page. 

Closes #47 

### Secondary
The integration unit test steps have been commented out in the runner (`.github/workflows/main_pr_tests.yaml`) for now after discussing with Bryson.

nstrauss commented on docs/user/getting_started.rst:
```diff
@@ -33,35 +33,98 @@ Create a Client
+    ...         server="jamf.my.org"    
+    ...     )    
+    ... )
 
```
Need another header here. Goes straight into access token expiration and client objects.

nstrauss reviewed (changes requested):
Since this is a breaking change, can the fact that it is be called out more? Whether in the docs or when trying to import `BasicAuthProvider`. Right now if someone were using the old class they'd only get an error like...

```
ImportError: cannot import name 'BasicAuthProvider' from 'jamf_pro_sdk.clients.auth'
```

When we cut a new release we'll include it there for sure as well. Most of my comments are around clarity and improving the docs, especially for newcomers. I tested the changes and everything works as expected. Let me know your thoughts on defaulting to `ApiClientCredentialsProvider` and if you have any questions.

liquidz00 commented:
All great callouts, I will definitely get these implemented. 

> Let me know your thoughts on defaulting to ApiClientCredentialsProvider and if you have any questions.

We are 100% aligned here. I also see basic auth as legacy. I'll be sure to sprinkle in some warning admonitions throughout that recommend using `ApiClientCredentialsProvider` over `UserCredentialsProvider`. I also like the idea of using the ApiClientCreds provider in usage examples. 

Any preference on how the breaking changes are called out in the docs? I'm personally partial to [Sphinx admonitions for describing changes between versions](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#describing-changes-between-versions), but am open to alternate approaches.

liquidz00 commented on docs/user/getting_started.rst:
```diff
@@ -31,37 +31,124 @@ When running ``pip freeze`` the SDK will appear with a filepath to the source in
-    >>> from jamf_pro_sdk import JamfProClient, BasicAuthProvider
+.. important::
+
+    **Breaking Change**: As of version ``0.8a1``, the SDK no longer uses ``BasicAuthProvider`` objects. Use :class:`~jamf_pro_sdk.clients.auth.ApiClientCredentialsProvider` as the new default.
```
I did my best guess as to what version the SDK would be bumped to. Just wanted to call out this line specifically in case I was way off.

jp-cpe commented:
Wow, this is great work @liquidz00! I will test these changes as well.

liquidz00 commented:
@jp-cpe thank you!

liquidz00 commented:
FYI: I added some logic to `load_from_keychain` to ensure the server URL is properly formed with `https://` if it was not passed, or if `http://` was used. In my scratch testing, this stopped the `CredentialError` from being thrown. 

```python
from src.jamf_pro_sdk import (
    JamfProClient,
    ApiClientCredentialsProvider,
    UserCredentialsProvider,
    load_from_keychain,
)

user_client = JamfProClient(
    server="anyorg.jamfcloud.com",
    credentials=load_from_keychain(UserCredentialsProvider, "anyorg.jamfcloud.com"),
)
print(user_client)

api_client = JamfProClient(
    server="anyorg.jamfcloud.com",
    credentials=load_from_keychain(ApiClientCredentialsProvider, "anyorg.jamfcloud.com"),
)
print(api_client)
```

I tested both scenarios--omitting the scheme entirely and only passing `http://` and received client objects as expected. I added some extra information in the 'important' admonition under the load from keychain header in the getting started docs.

nstrauss commented on src/jamf_pro_sdk/clients/auth.py:
```diff
@@ -227,95 +227,140 @@ def _request_access_token(self) -> AccessToken:
-        password = credentials["password"]
 
-        super().__init__(username, password)
+def load_from_keychain(
```
I looked at the existing class for this again (https://github.com/macadmins/jamf-pro-sdk-python/blob/cbb641ad3cd1ab82dc24d706ca053ccb88eb88c6/src/jamf_pro_sdk/clients/auth.py#L295-L321) and think changes could benefit from simplification.

`UserCredentialsProvider` prompts for username and only requires one keychain item. `ApiClientCredentialsProvider` doesn't prompt for client ID and requires two separate keychain items. It feels inconsistent. I think providing the non-secret item as a field is easier and more user friendly. 

Made example suggestions below. Then you could create a keychain item and use something like this without user interaction. If you like it and decide to go this route, may want to add additional validation for `client_id` vs `username` depending on provider, or move to a single common field like "identifier" or "identity". 

```python
client = JamfProClient(
    credentials=load_from_keychain(
        provider_type=ApiClientCredentialsProvider,
        server="my.jamfcloud.com",
        client_id="908A87FA-4BE7-4B06-81B7-893A30D499DA",
    ),
)
```

liquidz00 replied:
Didn't think about the inconsistency, thats a good callout. 

> If you like it and decide to go this route, may want to add additional validation for client_id vs username depending on provider, or move to a single common field like "identifier" or "identity".

The `identity` route isn't a bad idea either. My only thought on that is that passing an identity would imply `provider_type`. We could simplify things by removing the provider type parameter. Something to the effect of: 
```python
from typing import Literal

def load_from_keychain(
    server: str,
    identity: Literal["username", "client_id"] = "client_id",  # Default to ApiCredsProvider
    client_id: Optional[str] = None,
    username: Optional[str] = None
) -> CredentialsProvider:
    if identity == "username": 
        password = keyring.get_password(server, username)
        # Check for password being 'None' here
        return UserCredentialsProvider(username, password)
    elif identity == "client_id":
        password = keyring.get_password(server, client_id)
        # Check for password being 'None' here
        return ApiClientCredentialsProvider(client_id, password)
```

The repeated checks for password being None can probably be sharpened up in a more efficient manner, but curious as to your thoughts on this approach.

liquidz00 replied:
Actually, scratch that. In hindsight, I don't believe that would scale well if alternate Cred Providers are introduced in the future, or for folks using custom Cred Providers.

nstrauss commented on src/jamf_pro_sdk/clients/auth.py:
```diff
@@ -227,95 +227,140 @@ def _request_access_token(self) -> AccessToken:
 
-        super().__init__(username, password)
+def load_from_keychain(
+    provider_type: Type[CredentialsProvider], server: str
```
```suggestion
    provider_type: Type[CredentialsProvider],
    server: str,
    client_id: Optional[str] = None,
    username: Optional[str] = None
```

nstrauss commented on src/jamf_pro_sdk/clients/auth.py:
```diff
@@ -227,95 +227,140 @@ def _request_access_token(self) -> AccessToken:
 
+    if issubclass(provider_type, UserCredentialsProvider):
+        username = input("Jamf Pro Username: ")
+        password = keyring.get_password(service_name=server, username=username)
         if password is None:
             raise CredentialsError(
                 f"Password not found for server {server} and username {username}"
             )
-
-        super().__init__(username, password)
+        return provider_type(username, password)
+    elif issubclass(provider_type, ApiClientCredentialsProvider):
+        client_id = keyring.get_password(service_name=server, username="CLIENT_ID")
+        client_secret = keyring.get_password(service_name=server, username="CLIENT_SECRET")
+        if not client_id or not client_secret:
+            raise CredentialsError(
+                f"API Credentials were not found for server {server}. Please verify they are in the correct format."
+            )
+        return provider_type(client_id, client_secret)
+    else:
+        raise TypeError(f"Unsupported credentials provider: {provider_type}")
```
```python
if issubclass(provider_type, UserCredentialsProvider):
    identity = username
elif issubclass(provider_type, ApiClientCredentialsProvider):
    identity = client_id
else:
    raise TypeError(f"Unsupported credentials provider: {provider_type}")

password = keyring.get_password(service_name=server, username=identity)
if password is None:
    raise CredentialsError(
        f"Password not found for server {server} and username {username}"
    )

return provider_type(identity, password)
```
